src/core/net/dns_client.cpp

```diff
@@ -1106,6 +1191,121 @@ void Client::HandleTimer(void)
     }
 }
 
+#if OPENTHREAD_CONFIG_DNS_OVER_TCP_ENABLE
+void Client::HandleTcpTimer(void)
+{
+    // If message wasn't received in timeout time close connection.
+    mEndpoint.SendEndOfStream();
+    mTcpInitialized = false;
+}
+
+void Client::HandleTcpEstablished(otTcpEndpoint *aEndpoint)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    mEndpoint.SendByReference(mSendLink, 0);
+}
+
+void Client::HandleTcpEstablishedCallback(otTcpEndpoint *aEndpoint)
+{
+    LogDebg("Dns tcp connection established.\n");
+    static_cast<Client *>(otTcpEndpointGetContext(aEndpoint))->HandleTcpEstablished(aEndpoint);
+}
+
+void Client::HandleTcpSendDoneCallback(otTcpEndpoint *aEndpoint, otLinkedBuffer *aData)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aData);
+}
+
+void Client::HandleTcpReceiveAvailable(otTcpEndpoint *aEndpoint,
+                                       size_t         aBytesAvailable,
+                                       bool           aEndOfStream,
+                                       size_t         aBytesRemaining)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aBytesRemaining);
+
+    const otLinkedBuffer *data;
+    size_t                totalReceived = 0;
+    Message              *message       = nullptr;
+    Response              response;
+    QueryType             type;
+    Error                 responseError;
+    uint16_t              length = 0;
+
+    if (aEndOfStream)
+    {
+        // Cleanup is done in dissconnected callback.
+        mTcpTimer.Stop();
+    }
+
+    VerifyOrExit(aBytesAvailable != 0);
+    mEndpoint.ReceiveByReference(data);
+    message = mSocket.NewMessage(0);
+    VerifyOrExit(message != nullptr);
+
+    if (data != nullptr)
+    {
+        totalReceived += data->mLength;
+        message->AppendBytes(data->mData + kTcpQueryLenSize, data->mLength - kTcpQueryLenSize);
+        length = ReadUint16(data->mData);
+        if (length > OPENTHREAD_CONFIG_DNS_OVER_TCP_QUERY_MAX_SIZE)
+        {
+            LogWarn("Dns query over wasn't received message too big.");
+            responseError = kErrorNoBufs;
+        }
+    }
+    mEndpoint.CommitReceive(totalReceived, 0);
+    response.mInstance = &Get<Instance>();
```

Comment:
Here we are assuming we get the entire query response in one chunk (first invocation of `HandleTcpReceiveAvailable()` callback).
   - I think this may not be always the case and we may rx the query response in chunks.
   - We need to handle it.

Here is my suggestion:
   - The quey response, the first two bytes indicate the length of msg.
   - We can read the length and check if `aBytesAvailable` matches what we have received.
   - If not, we can let TCP keep the bytes, don't call `CommitReceive()` and wait for next `HandleTcpReceiveAvailable`
   - Only when we have all, we can allocate a `message` and read and `CommitReceive().

Reply:
Please check new way of handling it.

Reply:
I think we still need to update code to handle receiving response in chunks.

Suppose the query response from server is 100 bytes. We know server will append length as `uint16_t` at start. So overall, we expect to get 102 bytes over TCP.

`HandleTcpReceiveAvailable` is called as TCP receives chunks. So it can be called when we only get 20 bytes of data.
- In the code now, I think we read the 20 bytes and try to parse it as query response which will not be correct.
- We need to wait till we get all bytes. Here is my suggestion:
   - Read the first two bytes to determine the expected `length`
   - Check if `aBytesAvailable` is larger than what we expect (`length + sizeof(uint16)`).
   - If so read and then call `CommitReceive()` which lets `Tcp` know that we are done with the received content and it can be freed and/or reused).
   - Otherwise, we don't call `CommitReceive()` and wait for next call of `HandleTcpReceiveAvailable` (`Tcp` will keep the received content for us util we call `CommitReceive()`).